src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4110,26 +4111,43 @@ void MergeTreeData::backupData(BackupEntriesCollector & backup_entries_collector
     backup_entries_collector.addBackupEntries(backupParts(data_parts, data_path_in_backup));
 }
 
-BackupEntries MergeTreeData::backupParts(const DataPartsVector & data_parts, const String & data_path_in_backup)
+BackupEntries MergeTreeData::backupParts(const DataPartsVector & data_parts, const String & data_path_in_backup) const
 {
     BackupEntries backup_entries;
     std::map<DiskPtr, std::shared_ptr<TemporaryFileOnDisk>> temp_dirs;
 
+    /// Tables in atomic databases have UUID. When using atomic database we don't have to create hard links to make a backup, we can just
+    /// keep smart pointers to data parts instead. That's because the files of a data part are removed only by the destructor of the data part
```

Comment:
> we can just keep smart pointers to data parts instead. That's because the files of a data part are removed only by the destructor of the data part

It's not true, see `MergeTreeData::grabOldParts` and `MergeTreeData::clearPartsFromFilesystem`

Reply:
Ok, but then how can holding a StoragePtr help to protect a part's files from removing?